File: src/service/session/transactions/signup.rs
use crate::transaction::traits::{Body};
use crate::service::session::transactions::regex::{check_name, check_email, check_pwd};

use std::error::Error;
use std::any::Any;

pub struct TxSignup {
    name: String,
    addr: String,
    pwd: String,
}

impl TxSignup {
    pub fn new(name: String, addr: String, pwd: String) -> Self {
        TxSignup{
            name: name,
            addr: addr,
            pwd: pwd,
        }
    }
}

impl Body for TxSignup {
    fn precondition(&self) -> Result<(), String> {
        check_name(&self.name)?;
        check_email(&self.addr)?;
        check_pwd(&self.pwd)?;
        
        Ok(())
    }

	fn postcondition(&self) -> Option<Result<Box<dyn Any>, String>> {
        None
    }

	fn commit(&self) -> Result<(), Box<dyn Error>> {
        Ok(())
    }

	fn rollback(&self) {

    }

}

File: service/session/transactions/signup.go
package transactions

import (
	"context"
	"log"

	pb "github.com/alvidir/tp-auth/proto/session"
)

// TxSignup represents an
type TxSignup struct {
	req *pb.SignupRequest
}

// Precondition validates the transaction is ready to run
func (tx *TxSignup) Precondition() error {
	return nil
}

// Postcondition creates new user and a opens its first session
func (tx *TxSignup) Postcondition(context.Context) (interface{}, error) {
	log.Printf("Got a Signup request")
	return nil, nil
}

// Commit commits the transaction result
func (tx *TxSignup) Commit() error {
	return nil
}

// Rollback rollbacks any change caused while the transaction
func (tx *TxSignup) Rollback() {

}
